odoo/tools/mail.py

```diff
@@ -283,6 +297,31 @@ def is_html_empty(html_content):
     tag_re = re.compile(r'\<\s*\/?(?:p|div|span|br|b|i)(?:(?=\s+\w*)[^/>]*|\s*)/?\s*\>')
     return not bool(re.sub(tag_re, '', html_content).strip())
 
+def parse_small_qweb(text):
+    reg = re.compile(r"""
+    (
+        (?:
+            (?:\\\\)
+            |(?:\\\$)
+            |(?:(?!\$\{).)
+        )*
+    )
+    (?:\$\{
+        ((?:[^}\'\"\\]
+            |'(?:[^'\\]|\\.)*'?
+            |"(?:[^"\\]|\\.)*"?
+            |(?:\\.))*)
+    })?
+    """, re.X | re.DOTALL)
```

Comment:
this is prone to catastrophic backtracking, leading to deadlocking the system when a malformed or unexpected expression is parsed.

Example:
```
${hello "this is' "a 'nice "pathological 'expression "that 'may "deadlock
```
takes 30s+ to evaluate, and it's exponential, so you can make it take years to parse with a few more words..

Reply:
I haven't thought about the regex perfs. I've learned about atomic group and came up with a regex that seems to handle the pathological case you mention with 100.000+ character in less than a second.
Here is the new regex: 

```
    (?=(?P<a>
        (?:
            (?:\\\\)
            |(?:\\\$)
            |(?:(?!\$\{).)
        )*
    ))(?P=a)
    (?:\$\{
        (?=(?P<b>(?:
            [^}\'\"\\]
            |'
                (?=(?P<c>
                    (?:[^'\\]|\\.)*
                ))(?P=c)'?
            |"
                (?=(?P<d>
                    (?:[^"\\]|\\.)*
                ))(?P=d)"?
            |
                (?:\\.)
        )*))(?P=b)
    })?
```

@odony What do you think about it?